Mocks vignette (#155)
GSK-Biostatistics/tfrmt: vignettes/building_mocks.Rmd

---
title: "Building Mocks"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{Building Mocks}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
```

```{r message=FALSE}
library(tfrmt)

> library(dplyr) # for arrange_all()

Reviewer: ```suggestion
library(dplyr) 
```

> library(tidyr) # for crossing()

Reviewer: ```suggestion
library(tidyr) 
```

```

Mock tables are useful for visualising how future generated data may be 
presented in its final form. Using the `{tfrmt}` package you can create a mock 
table using placeholders (e.g. `xx.x`, `XXX`) or using mock data with numerical 

> values. Note that using the lower or upper case is up to your preference.

Author: included in latest commit

Tables are generally built by creating a tfrmt specification tailored to your 
data, and then feeding that into one of the available print functions. In this 

> article we will only explore the use of the `print_mock_gt()` function. For 

Reviewer: ```suggestion
article we will only explore the use of the `print_mock_gt()` function, which is specialized for printing mock tables to gt. For 
```

information on `print_to_gt()` please refer to its documentation.

All tfrmt specifications are created using the `tfrmt()` function. The 
foundation of this function are the `label`, `column`, `param` and `values` 
parameters. From here you can specify a `body_plan` where custom 
`frmt_structures` can be added and built upon.

> **Note:** layering has been used to avoid repeating code; for 

Reviewer: If possible, I think we should avoid the use of building template functions in this vignette. That is a more advanced topic while mocks are not.

more information please see the "Reuse and layer tfrmts" article.

## Making a simple mock table

In the following example only one format structure has been specified, where 
the desired outcome is:
<ul>

> <li> the count value followed by the percent value </li>

Reviewer: ```suggestion
<li> the count value and percent value are combined into the same cell </li>
<li> the count value followed by the percent value </li>
```

<li> the count value is formatted to be maximum three digits </li>
<li> the percent value is formatted to be two digits rounded to one decimal place </li>
</ul>

```{r}

> mock_tbl <- function(tfrmt_obj){

Reviewer: see comment above - Is this function necessary?

Author: Layering removed in recent update

  tfrmt(
    # Specify columns in the data
    label = label,
    column = column, 
    param = param,
    values = value,
    sorting_cols = c(ord1),
    
    # Specify body plan
    body_plan = body_plan(
      frmt_structure(group_val = ".default", label_val = ".default",
                     frmt_combine(
                       "{count} {percent}",
                       count = frmt("xxx"),
                       percent = frmt_when("==100"~ frmt(""),
                                           "==0"~ "",
                                           "TRUE" ~ frmt("(xx.x%)"))))
      ))
}
```

> The `print_mock_gt()` function has parameters `.default` and `n_cols` that 

Reviewer: First, we will explore how to create a mock table using this tfrmt object alone. The `print_mock_gt()` function has parameters `.default` and `n_cols`; these options inform the number of rows and columns of the mock table. In this example, we will use their default values (3) to print three row labels and three columns.

Author: Implemented change in latest commit

data you can utilise the `crossing()` function from the `{tidyr}` package. You 
may also wish to add an integer column specifying order layer for complex tables.

Here is an example:

```{r}
df <- crossing(label = c("label 1", "label 2", "label 3"),
               column = c("trt1", "trt2", "trt1&trt2", "pl"),
               param = c("count", "percent")) %>%
  # Assign numerical order (optional)
  mutate(ord1 = rep(seq(1:length(unique(.$label))), each = nrow(.)/length(unique(.$label)) ))

df
```

This time the `df` data has been added to the `print_mock_gt()` function, so 
now you can see that the label and column names have changed.

```{r}
# Print table
mock_tbl() %>%
  print_mock_gt(df)
```

You can adjust column names and structure using the `col_plan` parameter. It 
accepts selection helpers from the `{tidyselect}` package. For example, you can 
select `everything()` and remove columns that `starts_with("ord")`.

You can also use `span_structure()` to add spanning labels as below.

Note: You might see a warning along the lines of 
`restarting interrupted promise evaluation`.

```{r warning=FALSE}

> mock_tbl_adjust_cols <- function(){

Reviewer: again, i'd probably avoid the use of the function and just pipe everything together

Author: see above

  mock_tbl() %>%
  tfrmt(
    # Specify column structure
    col_plan = col_plan(
      -starts_with("ord"),
      span_structure(
        "Treatment",
        T1 = trt1,
        T2 = trt2,
        `T1&T2`= `trt1&trt2`
        ),
      span_structure(
        "Placebo",
        PL = pl
        ))
  ) 
}
  
mock_tbl_adjust_cols() %>%
  print_mock_gt(df)
```

Alternatively you can add a span column to your dataset and allocate names for 

> each column. Subsequently, ensure you concatenate your `span` and `column` 

Reviewer: ```suggestion
each column. Subsequently, ensure you include both your `span` and `column` 
```

Author: Implemented change in latest commit

to apply to the the new columns.

```{r}
tfrmt(
  # Specify columns in the data
  label = "label",
  param = "param",
  column = "column",
  values = value,
  
  # Specify body plan
  body_plan = body_plan(
    frmt_structure(group_val = ".default", label_val = ".default",
                   frmt_combine(
                     "{count} {percent}",
                     count = frmt("xxx"),
                     percent = frmt_when("==100"~ frmt(""),
                                         "==0"~ "",
                                         "TRUE" ~ frmt("(xx.x%)")))),
    frmt_structure(group_val = ".default", label_val = ".default",
                   frmt_combine(
                     "{num} ({lower}, {upper})",
                     num = frmt("xx.x"),
                     lower = frmt_when("==100"~ frmt(""),
                                       "==0"~ "",
                                       "TRUE" ~ frmt("xx.x%")),
                     upper = frmt_when("==100"~ frmt(""),
                                       "==0"~ "",
                                       "TRUE" ~ frmt("xx.x%"))))
    ),
  col_plan = col_plan(
    -starts_with("ord"))
  ) %>%
  print_mock_gt(df)
```

> Note that if you'd like to print a select few columns from your dataset you 

Reviewer: Should we just show how this can be done in the col_plan instead?

Author: Changed example to demonstrate selecting specific columns with `col_plan` but also kept note mentioning that `filter()` can be used on `df` within `print_mock_gt()` as an alternative

  label = "label",
  param = "param",
  column = "column",
  values = value,
  sorting_cols = vars(ord1),
  
  # Specify body plan
  body_plan = body_plan(
    frmt_structure(group_val = ".default", label_val = ".default",
                   frmt_combine(
                     "{count} {percent}",
                     count = frmt("xxx"),
                     percent = frmt_when("==100"~ frmt(""),
                                         "==0"~ "",
                                         "TRUE" ~ frmt("(xx.x%)")))),
    frmt_structure(group_val = ".default", label_val = ".default",
                   frmt_combine(
                     "{num} ({lower}, {upper})",
                     num = frmt("xx.x"),
                     lower = frmt_when("==100"~ frmt(""),
                                       "==0"~ "",
                                       "TRUE" ~ frmt("xx.x%")),
                     upper = frmt_when("==100"~ frmt(""),
                                       "==0"~ "",
                                       "TRUE" ~ frmt("xx.x%"))))
    ),
  col_plan = col_plan(
    -starts_with("ord"))
  ) %>%
  print_mock_gt(df %>% filter(!stringr::str_detect(column, "^Risk Diff")))
```

> ### Same value formatting but 1 grouping level

Author: see above
